### **MapReduce** Implementation - Ireflix Viewing Log Analysis

### ***Condition:
- Using Python, implement a simple MapReduce solution that processes the Ireflix viewing log below
- Create a sample dataset of at least 15 viewing records in the format (genre, watchtime_minutes) covering a minimum of 4 different genres.
- Write a mapper() function that reads each line and outputs (genre, 1) key-value pairs.
- Write a reducer() function that groups by genre and returns the total view count per genre.
- Write a main() function that calls both functions and prints the final output.
- Run your solution, provide a screenshot of the output, and in one sentence explain what the result tells Ireflix about their most watched genre.
### ***Solution:
Our MapReduce solution processes Ireflix viewing logs to analyze genre 
popularity based on real database data from CA1. The implementation 
processes sample (from movies.csv) viewing records covering 6 different genres: Action, 
Drama, Comedy, Horror, Sci-Fi, and Thriller.

In [ ]:
# Importing necessary libraries

from collections import defaultdict
import pandas as pd

 Using synthetic movie dataset generated for Ireflix database

In [ ]:
# INPUT DATA: Load CSV

data = pd.read_csv('../DATASETS/movies.csv')

In [ ]:
# MAPPER FUNCTION 

def map_function(data):
    mapped = []
    for genre in data ['genre']:
        mapped.append((genre, 1))   
    return mapped

In [ ]:
# SHUFFLE AND SORT.

def shuffle_sort(mapped_data):
    grouped_data = defaultdict(list)
    for key, value in mapped_data:
        grouped_data[key].append(value)
    return grouped_data
  

In [ ]:
# REDUCER FUNCTION.

def reduce_function(grouped_data):
    reduced_data = {}
    for key, values in grouped_data.items():
        reduced_data[key] = sum(values)
    return reduced_data

In [14]:
# EXECUTE MAPREDUCE

mapped_data = map_function(data)
grouped_data = shuffle_sort(mapped_data)
reduced_data = reduce_function(grouped_data)
print(reduced_data)
print(mapped_data)
print(grouped_data)

{'Sci-Fi': 162, 'Romance': 160, 'Comedy': 147, 'Action': 153, 'Horror': 164, 'Drama': 165, 'Thriller': 149}
[('Sci-Fi', 1), ('Romance', 1), ('Comedy', 1), ('Action', 1), ('Comedy', 1), ('Horror', 1), ('Romance', 1), ('Action', 1), ('Sci-Fi', 1), ('Drama', 1), ('Thriller', 1), ('Comedy', 1), ('Drama', 1), ('Horror', 1), ('Action', 1), ('Drama', 1), ('Horror', 1), ('Comedy', 1), ('Thriller', 1), ('Comedy', 1), ('Sci-Fi', 1), ('Drama', 1), ('Horror', 1), ('Thriller', 1), ('Sci-Fi', 1), ('Horror', 1), ('Action', 1), ('Thriller', 1), ('Action', 1), ('Sci-Fi', 1), ('Thriller', 1), ('Thriller', 1), ('Drama', 1), ('Comedy', 1), ('Sci-Fi', 1), ('Sci-Fi', 1), ('Comedy', 1), ('Sci-Fi', 1), ('Action', 1), ('Comedy', 1), ('Thriller', 1), ('Thriller', 1), ('Comedy', 1), ('Comedy', 1), ('Horror', 1), ('Action', 1), ('Sci-Fi', 1), ('Romance', 1), ('Comedy', 1), ('Romance', 1), ('Action', 1), ('Romance', 1), ('Horror', 1), ('Sci-Fi', 1), ('Comedy', 1), ('Thriller', 1), ('Sci-Fi', 1), ('Thriller', 1), (

In [13]:
print("== IREFLIX GENRES ANALYSIS ==")

for genre, total in reduced_data.items():
    print(genre, ":", total)

== IREFLIX GENRES ANALYSIS ==
Sci-Fi : 162
Romance : 160
Comedy : 147
Action : 153
Horror : 164
Drama : 165
Thriller : 149


The MapReduce analysis shows that Drama (165) and Horror(164) are among the most watched genres (for movies) on the Ireflix platform.